test/format/ops.c

```diff
@@ -1706,6 +1613,63 @@ row_insert(TINFO *tinfo, WT_CURSOR *cursor, bool positioned)
     return (0);
 }
 
+/*
+ * col_insert_resolve --
+ *     Resolve newly inserted records.
+ */
+static void
+col_insert_resolve(TINFO *tinfo)
+{
+    uint64_t v, *p;
+    u_int i;
+
+    /*
+     * We don't want to ignore column-store records we insert, which requires we update the "last
+     * row" so other threads consider them. Threads allocating record numbers can race with other
+     * threads, so the thread allocating record N may return after the thread allocating N + 1. We
+     * can't update a record before it's been inserted, and so we can't leave gaps when the count of
+     * records in the table is incremented.
+     *
+     * The solution is a per-thread array which contains an unsorted list of inserted records. If
+     * there are pending inserts, we review the table after every operation, trying to update the
+     * total rows. This is wasteful, but we want to give other threads immediate access to the row,
+     * ideally they'll collide with our insert before we resolve.
+     *
+     * Process the existing records and advance the last row count until we can't go further.
+     */
+    do {
+        WT_ORDERED_READ(v, g.rows);
+        for (i = 0, p = tinfo->insert_list; i < WT_ELEMENTS(tinfo->insert_list); ++i) {
+            if (*p == v + 1) {
+                testutil_assert(__wt_atomic_casv64(&g.rows, v, v + 1));
```

Comment:
Just to make sure I understand: At first I was surprised that we're dying if the atomic operation fails. But this thread should be the only one with an insert that == v + 1, correct? So the atomic CAS should actually never fail for this thread.

Reply:
Yes, that's correct.

We don't really need the atomic update because we're the only writer as you pointed out, but we do want a barrier to ensure the write gets done immediately, so I used the atomic.

Let me know if you see a simpler solution.